[CamModels] Add extractor

mars67857 opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests: _Searching for "cammodels" repo-wide yielded no code or issue results_
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8): _Yes, on Python 3.6. Do I need to run it on other versions too?_

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

#### Extractor Info

Adds a new extractor for CamModels.com streaming video. Golden path example for a live stream: 

```
$ python3 -m youtube_dl -F https://www.cammodels.com/cam/AutumnKnight/
[CamModels] AutumnKnight: Downloading webpage
[CamModels] AutumnKnight: Downloading JSON metadata
[info] Available formats for AutumnKnight:
format code  extension  resolution note
256          flv        256x144     160k video@ 111k, audio@ 49k
654          flv        654x368     782k video@ 700k, audio@ 82k
1280         flv        1280x720   1736k video@1657k, audio@ 79k (best)
```

#### Current Limitations

I left the test case `_TEST` commented out. If you uncomment it and run `python3 test/test_download.py TestDownload.test_CamModels`, it passes vacuously, similar to the (useless) test code in `chaturbate.py`. 

Suggestions on how to meaningfully end-to-end test a streaming service where streams may or may not exist at a given time would be appreciated!

dstftw commented:
Read
>[adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections

and fix all issues.

mars67857 commented:
@dstftw 

- Deleted the useless test
- Deleted the optional metafields (`ext` and `age_limit`)
- Made all regex searches non-fatal, with appropriate `ExtractorErrors` being thrown
- Added fallback logic for parsing out video links using regexes, in the event they change their JSON response schema

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url_or_request=url, video_id=video_id, headers=self._HEADERS)
```
No need to name positional arguments. Why touch headers?

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+
+class CamModelsIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?cammodels\.com/cam/(?P<id>\w+)'
+    _MANIFEST_URL = r'manifestUrlRoot=(?P<id>https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*))'
```
Each piece used only once must be moved to the place where it's actually used.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+
+    def _get_formats_from_manifest(self, manifest, video_id):
+        try:
+            rtmp_formats = manifest['formats']['mp4-rtmp']['encodings']
```
All formats should be extracted.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+        try:
+            rtmp_formats = manifest['formats']['mp4-rtmp']['encodings']
+            formats = []
+            for format in rtmp_formats:
```
Don't shadow built-in names.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+                formats.append({
+                    'ext': 'flv',
+                    'url': format.get('location'),
+                    'width': format.get('videoWidth'),
```
`int_or_none` for all int fields.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+                    'format_id': str(format.get('videoWidth'))
+                })
+        # If they change the JSON format, then fallback to parsing out RTMP links via regex.
+        except:
```
No bare except.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+            private = self._html_search_regex(pattern=self._PRIVATE, string=webpage, name='private show indicator', fatal=False)
+            if private:
+                raise ExtractorError(msg=self._PRIVATE_CONSOLE_ERROR, expected=True, video_id=video_id)
+            raise ExtractorError(msg=self._MANIFEST_URL_CONSOLE_ERROR, expected=False, video_id=video_id)
```
Code duplication. Prepare message and then throw only once.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+    def _get_manifest_url_from_webpage(self, video_id, webpage):
+        manifest_url_root = self._html_search_regex(pattern=self._MANIFEST_URL, string=webpage, name='manifest', fatal=False)
+        if not manifest_url_root:
+            offline = self._html_search_regex(pattern=self._OFFLINE, string=webpage, name='offline indicator', fatal=False)
```
`fatal=False` will print error wen failed. Instead it should be optional and not print any error.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+            offline = self._html_search_regex(pattern=self._OFFLINE, string=webpage, name='offline indicator', fatal=False)
+            if offline:
+                raise ExtractorError(msg=self._OFFLINE_CONSOLE_ERROR, expected=True, video_id=video_id)
+            private = self._html_search_regex(pattern=self._PRIVATE, string=webpage, name='private show indicator', fatal=False)
```
Same.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+        webpage = self._download_webpage(url_or_request=url, video_id=video_id, headers=self._HEADERS)
+        manifest_url = self._get_manifest_url_from_webpage(video_id=video_id, webpage=webpage)
+        manifest = self._get_manifest_from_manifest_url(manifest_url=manifest_url, video_id=video_id, webpage=webpage)
+        formats = self._get_formats_from_manifest(manifest=manifest, video_id=video_id)
```
All methods only used once should be explicitly inlined.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+    _OFFLINE_CONSOLE_ERROR = 'This user is currently offline, so nothing can be downloaded.'
+    _PRIVATE = r'(?P<id>I’m in a private show right now)'
+    _PRIVATE_CONSOLE_ERROR = 'This user is doing a private show, which requires payment. This extractor currently does not support private streams.'
+    _MANIFEST_CONSOLE_ERROR = 'Link to stream info was found, but we couldn\'t access it. This stream may require login.'
```
This is pointless message since you don't support login.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+    def _get_manifest_from_manifest_url(self, manifest_url, video_id, webpage):
+        manifest = self._download_json(url_or_request=manifest_url, video_id=video_id, headers=self._HEADERS, fatal=False)
+        if not manifest:
+            raise ExtractorError(msg=self._MANIFEST_CONSOLE_ERROR, expected=False, video_id=video_id)
```
`note` and `errnote` of `_download_json` instead.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,83 @@
+        return manifest_url
+
+    def _get_manifest_from_manifest_url(self, manifest_url, video_id, webpage):
+        manifest = self._download_json(url_or_request=manifest_url, video_id=video_id, headers=self._HEADERS, fatal=False)
```
This is fatal.

mars67857 commented:
@dstftw I think I've fixed your review suggestions? For the headers question, I added a code comment explaining that their servers reject requests when a common `User-Agent` isn't supplied.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,93 @@
+        webpage = self._download_webpage(
+            url,
+            video_id,
+            headers=self._HEADERS)
```
Avoid unnecessary verbosity. Collapse these 3 lines into one. Same for all other places.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,93 @@
+            video_id,
+            headers=self._HEADERS)
+        manifest_url_root = self._html_search_regex(
+            r'manifestUrlRoot=(?P<id>https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*))',
```
It's pointless to match URLs with strict regex, needless to say it's invalid due to `&` being delimiter in query. Match anything after `https?://` till closing quote or `&`.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,93 @@
+            webpage,
+            'manifest',
+            None,
+            False)
```
`default` and `fatal` are not used together. Prefix keyword arguments with their names.

dstftw commented on youtube_dl/extractor/cammodels.py:
```diff
@@ -0,0 +1,93 @@
+                        'format_id': fmtName + str(encoding.get('videoWidth'))
+                    })
+        # If they change the JSON format, then fallback to parsing out RTMP links via regex.
+        except KeyError:
```
If processing of single format encoding fails it will incorrectly fallback here. `try` section should be minimal.
